CLI: add option to output json title set info

jstebbins opened the pull request.

jstebbins committed:
CLI: add option to output json title set info

lisamelton commented:
@jstebbins This is a great idea! Here's the file (again) that you attached to the other bug:

[json_title_set.txt](https://github.com/HandBrake/HandBrake/files/1423520/json_title_set.txt)

lisamelton commented:
@jstebbins That sample file is really cool but is there any documentation or code you can point me at that describes the taxonomy of the JSON? I want to make sure it contains everything I need.

Also, I assume the contents are always in English?

Thanks!

jstebbins commented:
Only documentation is the code.  

The format is more the result of evolution than design.  Title dictionary and json could use a bit of cleaning up still.  For example, some of the elements are represented as integers which are just our internal enums and defines where they should be more meaningful strings.  If you see something that you want to use that uses HandBrake internal enums, let me know so it can be fixed first.

https://github.com/HandBrake/HandBrake/blob/c22e7ed3614ce541d1038d5d8eb0eb06431047b4/libhb/hb_json.c#L318
https://github.com/HandBrake/HandBrake/blob/c22e7ed3614ce541d1038d5d8eb0eb06431047b4/libhb/hb_json.c#L103

lisamelton commented:
@jstebbins Will do! I'll take a look at this closely over the next few days and let you know.

sr55 commented:
Might want to put a newline char after the last json tag.  Maybe a starting and ending flag to make it easier for scripts to detects?

jstebbins committed:
Add starting tag "JSON Title Set:" and trailing newline

Makes json easier for scripts to find and more readable

jstebbins commented:
Starting flag makes sense to designate that the title set follows (as opposed to some other thing enclosed by {}).  I don't think an ending flag is necessary since the whole thing is contained in {}

sr55 commented:
hmm, Wonder if we should add this to help?

I was also thinking, not necessarily now, but encode progress output and end state being printed in json too?

sr55 commented:
probably --version too to make API version easily detectable.

lisamelton commented:
@sr55 Yes, please indicate this capability in the output of `--help`. The reason I ask is that determining capabilities using the output of `--version` is sometimes problematic for development and custom builds. But I can easily parse the output of `--help` if a new option is listed. Plus, this is useful for ordinary CLI users.

Also, have you considered a way of making the output of `--scan` as only JSON? This could obviate the need for start and end markers and ease parsing by standard tools by allowing unmodified piping of `HandBrakeCLI` output to another tool's input.

For example, you could add an argument to the `--scan` argument called "json" so that a user could simply invoke `--scan=json` to get only the easily parseable information.

sr55 commented:
A low verbosity mode was something that was on my CLI re-design list. (When that'll actually happen I don't know. It's been on my todo list for about 2 years now :( )

It's slightly complicated by the fact that some of the output its done by 3rd party libraries so that all has to be redirected manually.

What are you using --help to detect for?   That would be another one that's probably at risk of change.

lisamelton commented:
@sr55 I currently parse the output of `HandBrakeCLI --help` to determine two things:

1. Whether to use the `ca_aac` or `av_aac` encoder as the default for AAC output.

2. Whether to use the newer `auto-anamorphic` or older `strict-anamorphic` option name.

If there was another reliable way to do either test, I would be doing that.

Just to be clear, checking the output of `--version` is not reliable for development and custom platform builds.

sr55 commented:
ca_aac is Mac only, so you can detect OS at the script level for that.  ca_aac is default on mac.

auto-anamorphic is trickier if you include nightly builds.

lisamelton commented:
@sr55 Detecting the OS platform on Ruby is more complicated than you think. :) And I don't want to include a dependency on another Gem just to do that.

sr55 commented:
Then you can get it from the version info.  Pretty sure we include it on there, if not it's trivial to add if we add json output for it.

We probably want a get capabilities method to show the OS specific features are enabled.

ArtificialMarty commented:
Folks, this is just great!  I'd ask for 2 enhancements:

* decode the enums for the audio channellayout and codec;
* add an audio item per track that indicates "director commentary" vs. "main";
* for subtitles, there's something that identifies closed captioning vs. whatever; please enumerate (decoded) these.

With that info, I'd be prepared to retranscode my entire collection (32TB (and growing) of isos).

lisamelton commented:
@sr55 Yes, I can get the platform from the output of `--version`. But I currently parse the output of `--help` because then that single piece of code works for versions of `HandBrakeCLI` prior to 1.0 which don't have the `--version` option. :)

jstebbins commented:
Reducing verbosity of output is problematic.  Much of the output we use for diagnosing problems would have to be squelched.  And many of the libraries we use don't even have a way to limit verbosity.  My gut reaction is that we shouldn't mess with this.  Just intermix json with appropriate parsable tags in the output.

Progress can easily be made json since we have a function in libhb for generating this already.  

Version info would be relatively easy to add.  There is a variety of potentially useful info we could throw into version.  There are a number of things in build/libhb/project.h that would be nice to have in extended version information.

I'm wondering how we want to handle audio channel layout.  I guess exposing the defines for this is easiest. Since channel layout can be pretty much any permutation of individual channels, it's not practical to try to express this in human readable format.

We can expose audio type (and subtitle type) in the title json.  I'll make those strings with meaningful values.

CC vs. VOBSUB, etc is "Source" in the subtitle list items.  I would rather make that human readable text than have scrapers using the internal enum values.  These are prone to change since they are not part of any public API. But need to get feedback from @sr55 first.  I'm pretty sure changing that will require changes in the Windows GUI.

sr55 commented:
I used ChannelLayout int but only as a passthru item to other API's

I don't use Subtitle "Format" tag.
Source I translate from int to internal C# representation of subtitle type.It's a simple 0 -7 integer. Any change requires a trivial change on my end.

jstebbins commented:
@sr55 OK.  I'll propose a change for "Source" by updating this PR. It'll be something simple "CC608" instead of 3.

sr55 commented:
Fine by me. I'll adapt as soon as you checkin.

lisamelton commented:
Thinking about it now, I'm not sure whether this is the proper issue to include _all_ of my feedback below. Maybe I should open a separate issue or issues? That's because this issue seems to be focused on just adding the option to output title set information in JSON format. Let me know where I should be including this if not here.

Let me repeat that I think this feature is a great idea! My feedback is about the structure and formating of the title set information itself. Specifically, what's missing, what's ambiguous and what could be improved.

Of course, after more reflection and testing, I'm certain to have additional feedback. :)

Anyway, maybe I'm missing it but the sample provided by @jstebbins doesn't include whether HandBrake thinks the title set is derived from a mounted disc, a single ISO image, a disc image directory, or a single media file. And if it's a single media file, whether the format of that file Matroska (MKV), MP4, AVI, etc.

All this information is necessary. I can't just assume what HandBrake will do ahead of time based on whether the input is a directory or file and what file extension it has.

Also missing from the sample are any subtitle tracks which have the "forced" flag set. I assume that's because the sample is derived from a disc and it's problematic getting that information?

But for MKV file inputs, it's essential to know whether the "forced" flag is set for a subtitle. And whether the "default" flag is set for a subtitle in MP4 file inputs.

In fact, knowing the attributes for video, audio and subtitle tracks, i.e. which are "forced" and/or "default," is necessary. Otherwise, I have to keep parsing the whacky-ass stream information from Libav to figure that out.

It appears, as @ArtificialMarty suggested and @sr55 and @jstebbins discussed, you're already moving towards replacing the internal enumumeration values for `"BitRate":`, `"ChannelLayout":`, `"Codec":`, `"Format":`, `"Source":`, `"Type":`, etc. with actual meaningful names. That's great! Please leave all the opacity behind. Parsing individual attributes is _much_ easier than parsing meta attributes like `"Description":` and `"Language":`.

BTW, I'm confused as to your criteria for deciding between parentheses and brackets to offset elements within `"Language":` and other attributes. IMO, switching between the two doesn't make the text more readable. Just pick one and use that, otherwise it seems like you're trying to emphasize a difference that may not be relevant.

More importantly, I would think that brackets (and braces) should be avoided in JSON since they're syntactically relevant to the format. I would stick with parentheses in your descriptions for offsetting elements. Better yet, I suggest you use "; " to separate elements within these meta attributes as this would make them easier to parse, if necessary, and a semicolon isn't syntactically relevant in JSON.

Does the `"Index":` attribute map to how I would access a particular track with the API? If not, then I need something that indicates what value I need to pass to your `--audio` and `--subtitle` options for that track. Otherwise I can't use all this nifty JSON when I call `HandBrakeCLI` again.

I'm impressed the `"FrameRate":` is so precise with a separate numerator and denominator, but I need something simple like `23.976` or `29.97` too so I can write a simple switch statement on the data. Please include that.

I assume the "Ticks": attribute in `"Duration":` is a total of the hours, minutes and seconds? If so, consider renaming it `"TotalTicks":` to indicate that it's a summation unlike the other three.

Again, more feedback will come. Thanks so much for asking! You folks are great.

jstebbins commented:
I added subtitle attributes, audio attributes, progress and version json output.

I had a look at audio and subtitle type and there are more possible permutations than I expected.  So I think it's best to keep these as int values. But I changed them to bit fields to capture all the possible permutations.  I've added defines for these to common.h
```
#define HB_SUBTITLE_ATTR_UNKNOWN    0x0000
#define HB_SUBTITLE_ATTR_NORMAL     0x0001
#define HB_SUBTITLE_ATTR_LARGE      0x0002
#define HB_SUBTITLE_ATTR_CHILDREN   0x0004
#define HB_SUBTITLE_ATTR_CC         0x0008
#define HB_SUBTITLE_ATTR_FORCED     0x0010
#define HB_SUBTITLE_ATTR_COMMENTARY 0x0020
#define HB_SUBTITLE_ATTR_4_3        0x0040
#define HB_SUBTITLE_ATTR_WIDE       0x0080
#define HB_SUBTITLE_ATTR_LETTERBOX  0x0100
#define HB_SUBTITLE_ATTR_PANSCAN    0x0200

#define HB_AUDIO_ATTR_NONE              0x00
#define HB_AUDIO_ATTR_NORMAL            0x01
#define HB_AUDIO_ATTR_VISUALLY_IMPAIRED 0x02
#define HB_AUDIO_ATTR_COMMENTARY        0x04
#define HB_AUDIO_ATTR_ALT_COMMENTARY    0x08
#define HB_AUDIO_ATTR_SECONDARY         0x10
```

I added a new element to the json called "Attributes" and it can take any bit-wise or of the above for audio or subtitles.

I changed the subtitle "Source" element to a string that can have the following values,"VOBSUB", "SRT", "CC608", "CC708", "UTF-8", "TX3G", "SSA", "PGS".

And for progress info, I changed the "State" element to a string in the json returned by hb_get_state_json().  The values can be:  "IDLE", "SCANNING", "SCANDONE", "WORKING", "PAUSED", "SEARCHING", "WORKDONE", "MUXING", "UNKNOWN".

jstebbins committed:
Add audio/sub attributes, version, and progress

jstebbins commented:
> Anyway, maybe I'm missing it but the sample provided by @jstebbins doesn't include whether HandBrake thinks the title set is derived from a mounted disc, a single ISO image, a disc image directory, or a single media file. And if it's a single media file, whether the format of that file Matroska (MKV), MP4, AVI, etc.

You didn't have this information before.  Why do you need it now?  None of the HandBrake front-end GUIs or CLI need to know this and therefore, it is not currently exposed.

> Also missing from the sample are any subtitle tracks which have the "forced" flag set. I assume that's because the sample is derived from a disc and it's problematic getting that information?

Although there is a "Forced" attribute that a DVD can set on a subtitle track, this does not necessarily indicate that the track has forced subtitles.  The only way to know this is to perform a subtitle scan of the entire disc searching for the "forced" flag that can be set on each individual subtitle in the track.  For other container formats that support forced or default flags, we can add this to subtitle attributes. 

> In fact, knowing the attributes for video, audio and subtitle tracks, i.e. which are "forced" and/or "default," is necessary.

HandBrake doesn't track these flags for audio or video.

> It appears, as @ArtificialMarty suggested and @sr55 and @jstebbins discussed, you're already moving towards replacing the internal enumumeration values for "BitRate":, "ChannelLayout":, "Codec":, "Format":, "Source":, "Type":, etc. with actual meaningful names.

BitRate is not an enumeration and needs no modification. As discussed already, ChannelLayout will remain an integer since it can be any permutation of individual channels which are represented as bit-fields in the integer. We use libav's AV_CH_LAYOUT_* defines for this.  You'll have to look up the values and parse the bit-fields appropriately if you want this info. Codec is a good candidate for making into a string. I'm not sure you need Format.  It only says whether the subtitles are text or bitmaps.  Source I already changed to a string and "Type" I added as "Attributes" which is another bit-field whose values are listed in my previous post..

> BTW, I'm confused as to your criteria for deciding between parentheses and brackets to offset elements within "Language"

There is no logic.  This field is for display purposes.  The formatting is what we deemed looked good in the GUI and CLI.

> More importantly, I would think that brackets (and braces) should be avoided in JSON since they're syntactically relevant to the format.

Irrelevant.  Strings are properly quoted in the json.

> Does the "Index": attribute map to how I would access a particular track with the API?

When reading a source that has multiple titles, some titles may be skipped for various reasons when building the list the user sees (e.g. title is too short, file is not a video file, etc...).  The "Index" records the position in the underlying source that the title was found.  I.e. it may be the 2nd entry the user sees, but it was the 5th entry in the DVD index structure.  It allows HandBrake to index to the correct element in the DVD (or other underlying source) structure.  In the title set, the value of "MainFeature" is an "Index".  So if you want to find the title that HandBrake detected as the main feature of a DVD, you have to find the title whose Index == MainFeature.

> I'm impressed the "FrameRate": is so precise with a separate numerator and denominator, but I need something simple like 23.976 or 29.97 too so I can write a simple switch statement on the data.

Not possible.  FrameRate can be *any* value.  Not just some small set of values that could be represented as you request.  If you want to confine your frontend to some set of values, you'll have to compute this yourself and find the closest match to the values you choose to support.

>I assume the "Ticks": attribute in "Duration": is a total of the hours, minutes and seconds?

A tick is 1/90000 second

jstebbins committed:
Add avformat forced/default flag to subtitle attributes

lisamelton commented:
@jstebbins Thanks so much for the thorough and clear reply!

>> Anyway, maybe I'm missing it but the sample provided by @jstebbins doesn't include whether HandBrake thinks the title set is derived from a mounted disc, a single ISO image, a disc image directory, or a single media file. And if it's a single media file, whether the format of that file Matroska (MKV), MP4, AVI, etc.

> You didn't have this information before. Why do you need it now? None of the HandBrake front-end GUIs or CLI need to know this and therefore, it is not currently exposed.

You're correct that there wasn't a way to disambiguate between ISO images and various directories formats as input. Wishful thinking on my part, I suppose. :) My apologies.

However it's easy to tell if the input was a single file and what format that was. This is included in the Libav stream information, e.g. `Scanning title 1 of 1, 0.00 %Input #0, matroska,webm, from ...` and such. My code, particularly that for the `convert-video` tool, requires this information.

I just assumed that HandBrake also tracked this. If you don't, then I'm going to continue needing the Libav stream information to parse. Not only for determine the input file format, but also for other metadata (see below).

>> Also missing from the sample are any subtitle tracks which have the "forced" flag set. I assume that's because the sample is derived from a disc and it's problematic getting that information?

> Although there is a "Forced" attribute that a DVD can set on a subtitle track, this does not necessarily indicate that the track has forced subtitles. The only way to know this is to perform a subtitle scan of the entire disc searching for the "forced" flag that can be set on each individual subtitle in the track. For other container formats that support forced or default flags, we can add this to subtitle attributes.

Yeah, that's what I meant, actually. :) My apologies for being unclear. I realize that determining "forced" for a track in a multi-title disc image is problematic without scanning.

But knowing "forced" and "default" for MKV and MP4 subtitle tracks is the important thing. And thanks so much for adding this!

>> In fact, knowing the attributes for video, audio and subtitle tracks, i.e. which are "forced" and/or "default," is necessary.

> HandBrake doesn't track these flags for audio or video.

I can probably live without the information for video tracks, but for audio I'm getting this from the Libav stream information now. That's how I know to verify the attributes are properly set when I switch the order of main audio tracks with the `convert-video` tool, e.g. when the AC-3 track is moved "in front" of the AAC track on conversion from MP4 to MKV.

And, like the file format information mentioned above, I can continue getting that from the Libav stream information. Please keep that information then.

>> It appears, as @ArtificialMarty suggested and @sr55 and @jstebbins discussed, you're already moving towards replacing the internal enumumeration values for `"BitRate":`, `"ChannelLayout"`:, `"Codec":`, `"Format":`, `"Source":`, `"Type":`, etc. with actual meaningful names.

> BitRate is not an enumeration and needs no modification.

Sorry, that was a stupid copy-and-paste error on my part including `"BitRate":` in there. :)

> As discussed already, ChannelLayout will remain an integer since it can be any permutation of individual channels which are represented as bit-fields in the integer. We use libav's AV_CH_LAYOUT_* defines for this. You'll have to look up the values and parse the bit-fields appropriately if you want this info.

Isn't the the point of JSON is human-readable text? That includes the data elements, too. Something like "5.1" is much more friendly and easier with which to do quick math. Oddly, I can get such formatting from parsing the Libav stream information.

> Codec is a good candidate for making into a string.

Agreed, and thank you!

> I'm not sure you need Format. It only says whether the subtitles are text or bitmaps.

Well, I have "Format" now and I use it. And it's really handy, too. Otherwise I'll have to assume that I know all the formats. While that's possible in the short term, it's prone to subtle breakage in the future.

> Source I already changed to a string and "Type" I added as "Attributes" which is another bit-field whose values are listed in my previous post..

OK, I missed that. Sorry.

>> BTW, I'm confused as to your criteria for deciding between parentheses and brackets to offset elements within "Language"

> There is no logic. This field is for display purposes. The formatting is what we deemed looked good in the GUI and CLI.

An honest man! :) OK, fair enough.

>> More importantly, I would think that brackets (and braces) should be avoided in JSON since they're syntactically relevant to the format.

> Irrelevant. Strings are properly quoted in the json.

Not completely. They're still a nuisance if you decide to parse them. :)

>> Does the "Index": attribute map to how I would access a particular track with the API?

> When reading a source that has multiple titles, some titles may be skipped for various reasons when building the list the user sees (e.g. title is too short, file is not a video file, etc...). The "Index" records the position in the underlying source that the title was found. I.e. it may be the 2nd entry the user sees, but it was the 5th entry in the DVD index structure. It allows HandBrake to index to the correct element in the DVD (or other underlying source) structure. In the title set, the value of "MainFeature" is an "Index". So if you want to find the title that HandBrake detected as the main feature of a DVD, you have to find the title whose Index == MainFeature.

Hmmm, I think we might be talking about two different things. You mean the titles. I mean the tracks within titles. To repeat, I need to know which track number `HandBrakeCLI` will use to access that track later. Or am I guaranteed that the order of tracks for a particular media type (i.e. audio or subtitle) within the JSON can be used to derive this information?

>> I'm impressed the "FrameRate": is so precise with a separate numerator and denominator, but I need something simple like 23.976 or 29.97 too so I can write a simple switch statement on the data.

> Not possible. FrameRate can be any value. Not just some small set of values that could be represented as you request. If you want to confine your frontend to some set of values, you'll have to compute this yourself and find the closest match to the values you choose to support.

Ah, I see what you mean now. That makes sense. Thanks!

>> I assume the "Ticks": attribute in "Duration": is a total of the hours, minutes and seconds?

> A tick is 1/90000 second

So it's not a total then. :)

jstebbins commented:
Input file format and whether the source is a single title or multiple titles is irrelevant to HandBrake.  The properties of each title is all that we care about.  The container does not affect those properties.

> I can probably live without the information for video tracks, but for audio I'm getting this from the Libav stream information now.

Adding this to the audio attributes for files parsed with libav is easy (and I'll add this).  But for DVD and BD, there is no default or forced audio flag.  For DVD and BD, the default audio (and subtitle for that matter) is selected by the menu system code that runs in the DVD or BD virtual machine.  So figuring out what track it is would require guessing what track the VM would select if it ran the menu code.  HandBrake is only capable of handling one video track, so it's always default :p

> Isn't the the point of JSON is human-readable text? That includes the data elements, too. Something like "5.1" is much more friendly and easier with which to do quick math. Oddly, I can get such formatting from parsing the Libav stream information.

First, I don't want you to get the idea I'm resisting changes here.  This topic deserves a long discussion because of the complexities involved.  So bare with me.

The point of json is to provide a *portable* text representation that is easily *machine* readable. It's the "portable" part that's tricky here. To make this truly portable, I would make ChannelLayout an array of individual channel names instead of a bit-field representing individual channels. But that would make things more complex for an application that wants to use ChannelLayout.  The application would have to translate this array of channel names back into some internal format that is more usable. So for the sake of keeping code simpler, ChannelLayout isn't as portable as it could be. The ChannelLayout field is meant to be as accurate a representation as is available to us of the channel mapping in the track. It can't be simplified to "5.1". 

But I see the potential need to simplify this.  The problem you face is that you don't have access to the libav functions that can tell you more detail about what the ChannelLayout represents.  But there are complexities to this problem that need to be explored in order to come up with the right option for simplification. You should to look at libavutil/channel_layout.h at AV_CH_* and how these are used.  libav recognizes 25 distinct types of channels.  "ChannelLayout" can be any combination of these 25 distinct channels. That means there are 2^25 (33554432) possible ChannelLayouts that we could get from libav. There is a set of commonly used channel layouts that *could* be represented by a set of strings (libav lists 28 "common" variations it recognizes).  But what we get from libav often doesn't neatly fall into one of these "common" layouts. We could supply a rough approximation of the ChannelLayout as a separate element.  For example, if the ChannelLayout exactly matches one of libav's 28 "common" layouts (libav has a function for this av_get_standard_channel_layout) we could map each of those to strings. If no exact match is found we could generate something like "\<non-lfe-count\>.\<lfe-count\>". But I'm not sure that's all that much better for you.  E.g. what are you going to do with "quad(side)" vs. "4.0"?  These are two different 4 channel, no lfe layouts.  You can find the complete list of strings I'm suggesting in libavutil/channel_layout.c search for channel_layout_map.  If you are trying to make "best" track selection choices based on layout, maybe the better thing would just be something like "ChannelCount" and "LFECount" elements.  It's not entirely clear to me what you want to use the "ChannelLayout" element for.

> Hmmm, I think we might be talking about two different things....

The only place there is an "Index" in the json is in the Title. Audio and subtitles are indexed by their position in their respective lists (starting with 1 since this is a CLI parameter).

> So it's not a total then. :)

"Ticks" *is* a total.  "Ticks" is the total duration represented as the number of 1/90000 second ticks.  So if the title's total duration is 1 minute, "Ticks" is 60 * 90000 = 5400000.  Same applies for chapters. If you want duration in finer detail than 1 second, you have to use ticks to calculate it.  Hours, Minutes, and Seconds fields are only used by the frontends for display purposes and are simply computed from Ticks.  TBH, I feel like Hours, Minutes, and Seconds shouldn't be there and libhb should just provide a helper function for computing these.  But legacy  :neutral_face:.  They don't hurt anyone.

jstebbins committed:
Fix dropped subtitle element

Didn't update json format spec when adding new element

lisamelton commented:
@jstebbins Again, thanks for the detailed reply!

> Adding this to the audio attributes for files parsed with libav is easy (and I'll add this).

Sweet! Thank you. That will be so much easier than parsing the Libav spew.

> First, I don't want you to get the idea I'm resisting changes here. This topic deserves a long discussion because of the complexities involved. So bare with me.

Done! And I see your point now.

> But I see the potential need to simplify this. The problem you face is that you don't have access to the libav functions that can tell you more detail about what the ChannelLayout represents. ...

That's part of the problem, yes.

> ... But there are complexities to this problem that need to be explored in order to come up with the right option for simplification. You should to look at libavutil/channel_layout.h at AV_CH_* and how these are used. libav recognizes 25 distinct types of channels. "ChannelLayout" can be any combination of these 25 distinct channels. That means there are 2^25 (33554432) possible ChannelLayouts that we could get from libav. ...

Woof! I had no idea is was that complicated. Thanks for the explanation.

>... There is a set of commonly used channel layouts that could be represented by a set of strings (libav lists 28 "common" variations it recognizes). But what we get from libav often doesn't neatly fall into one of these "common" layouts. We could supply a rough approximation of the ChannelLayout as a separate element. For example, if the ChannelLayout exactly matches one of libav's 28 "common" layouts (libav has a function for this av_get_standard_channel_layout) we could map each of those to strings. If no exact match is found we could generate something like "<non-lfe-count>.<lfe-count>".

That's actually what I was looking for! :) And I think Libav does this in their own console spew.

> But I'm not sure that's all that much better for you. E.g. what are you going to do with "quad(side)" vs. "4.0"? These are two different 4 channel, no lfe layouts. You can find the complete list of strings I'm suggesting in libavutil/channel_layout.c search for channel_layout_map. If you are trying to make "best" track selection choices based on layout, maybe the better thing would just be something like "ChannelCount" and "LFECount" elements.

Actually , having this "ChannelCount. LFECount" as a secondary simplified string would be even better. Yes, please do this. It's valuable even if it seems like duplicate information. (Much like Hours, Minutes and Seconds in the Duration field.)

> It's not entirely clear to me what you want to use the "ChannelLayout" element for.

The behavior of both `transcode-video` and `convert-video` changes significantly depending on whether tracks contain surround audio. This is a big feature, actually. And in the future when you support more than 5.1 channels for E-AC-3 encoding, then this will get even more sophisticated.

Basically, if the main audio track (or any other track designated to behave like the main audio track) has more than two channels or LFE, then I make sure that track is output twice, once in surround format and once in stereo format.

> The only place there is an "Index" in the json is in the Title. Audio and subtitles are indexed by their position in their respective lists (starting with 1 since this is a CLI parameter).

Great! Although it wouldn't hurt for the tracks to contain redundant index attributes since not all JSON parsing implementations keep array elements in their original order.

jstebbins committed:
set "default" audio attribute when available.

jstebbins commented:
> Actually , having this "ChannelCount. LFECount" as a secondary simplified string would be even better. 

These I would do as 2 separate integer elements.

> Great! Although it wouldn't hurt for the tracks to contain redundant index attributes since not all JSON parsing implementations keep array elements in their original order.

Any json parser that does not maintain array order is broken.  Maybe you are thinking of dictionary order?  The order of elements in a dictionary is not guaranteed.  But array order is.

lisamelton commented:
> These I would do as 2 separate integer elements.

That's an even better idea, now that I think about it! Thanks.

> Any json parser that does not maintain array order is broken. Maybe you are thinking of dictionary order? The order of elements in a dictionary is not guaranteed. But array order is.

Yeah, it was probably dictionary output where I saw variation before. Still, a redundant index isn't all that bad of an idea.

jstebbins committed:
make title json audio "Codec" a string

jstebbins committed:
add title json audio layout name, channel and lfe count

jstebbins commented:
Updated sample output
[json_title_set.txt](https://github.com/HandBrake/HandBrake/files/1428940/json_title_set.txt)

jstebbins committed:
make subtitle "Format" a string, "text" or "bitmap"

ArtificialMarty commented:
Thanks!  I think the latest changes cover everything I wanted!  Great work!

jstebbins commented:
I'm wondering if we should output the version information unconditionally when --json option is given.  This way, a scraper doesn't have to run HandBrake twice to discover the version.

lisamelton commented:
@jstebbins That's an excellent idea! Yes, please include the version and the platform.

jstebbins committed:
When --json enabled, always show version info

And add --json to help

jstebbins commented:
Sample output for a scan and an encode
[json_title_set.txt](https://github.com/HandBrake/HandBrake/files/1429123/json_title_set.txt)
[json_encode.txt](https://github.com/HandBrake/HandBrake/files/1429124/json_encode.txt)

jstebbins committed:
use stdout for json messages

jstebbins committed:
fix audio attributes for libav streams

they got overwritten after setting

ArtificialMarty commented:
I just had a thought: for the audio and subtitle attributes, would it be unreasonable to display them as an array of bit names?  Something like:

"attributes" : [ "normal", "forced", "cc608" ]

If nothing else, it would make them a lot more human-readable.

jstebbins commented:
That may be a fairly inconvenient representation for the other frontends.  In 'C' it's a lot easier to check if a bit is set:
```
if (attribute & FORCED)
  blah;
```
than it is to parse a list:
```
forced = 0
for (i = 0; i < count; i++)
  if (!strcmp(attribute[i], "forced") {
    forced = 1;
    break;
  }
if (forced)
  blah;
```

But lets get some feedback from the Window GUI developer.  He's the main user of the JSON APIs.  If C# has some kind of "contains" function for json arrays then he could do something like this which is functionally just as easy:
```
if (attribute contains "forced")
  blah
```

bradleysepos commented:
I would probably write a utility function to loop over JSON arrays and make native ones. Then higher level functions can be used.

ArtificialMarty commented:
I don't know C#, but in C:

int contains(char** stringarray, int n, char* tomatch)
{
        for (int i = 0; i < n; ++i)
        {
                if (strcmp(stringarray[i], tomatch) == 0)
                        return 1;
        }
        return 0;
}

// ...

if (contains(attributes, sizeof attributes / sizeof attributes[0], "forced"))
        blah;

Yeah, sizeof is icky, but C# may have an easier way.

jstebbins commented:
If we're going to change this, I think it should be a dictionary of boolean values, not an array.  order doesn't matter.

jstebbins committed:
make audio/subtitle attributes a dictionary of bools

jstebbins commented:
Sample
[json_title_set.txt](https://github.com/HandBrake/HandBrake/files/1431921/json_title_set.txt)

ArtificialMarty commented:
Thank you!  I now have a great reason to graduate from 1.0.7!  :)  And I get to toss my scan-scraper (but have to rewrite my transcode-a-disc Perl script, which I will now happily do since it will now finally have access to all the info it needs!)!  Yay team!

jstebbins committed:
revert Source and Codec to int, add SourceName and CodecName

jstebbins commented:
To simplify things for the Windows GUI, I've reverted audio Codec and subtitle Source to int values and added CodecName and SourceName as the string value equivalents.
[json_title_set.txt](https://github.com/HandBrake/HandBrake/files/1447214/json_title_set.txt)

sr55 commented:
Good to go from the Windows UI side. As soon  as you commit i'll push the required changes.

sr55 commented:
All Merged.

Thanks for the feedback all!
